KeyError: '__wrapped__' with from xxx import yyy

# Checklist

- [X] Include a detailed description of the bug or suggestion
- [ ] `pip list` of the virtual environment you are using
- [X] pytest and operating system versions
- [X] Minimal example if possible

# Description

I implemented `tox` and `pytest` for automated testing.  When I run the [MWE](https://github.com/ZaydH/pytest_bug), I get a `KeyError: '__wrapped__'` when I perform a `from xxx import yyy`.  

In the example I included, I do not get the KeyError if I comment out the line: 

...
from sty import fg
...


# Pip List

No virtual environment used.

# Version Info

*Python*: 3.6.5 & 3.7.1
*PyTest*: 4.4.0
*OS*: MacOS Mojave 10.14.4

# MWE

I created a very simple [repo](https://github.com/ZaydH/pytest_bug).  If I clone the repo and call either `tox` or `pytest` in the root directory, I get the error.

# Example Error Message

```
➜  pytest_bug git:(master) ✗ tox
GLOB sdist-make: /Users/zayd/repos/pytest_bug/setup.py
py36 recreate: /Users/zayd/repos/pytest_bug/.tox/py36
py36 installdeps: pytest, sty
py36 inst: /Users/zayd/repos/pytest_bug/.tox/.tmp/package/1/stratego-0.0.0.zip
py36 installed: atomicwrites==1.3.0,attrs==19.1.0,more-itertools==7.0.0,pluggy==0.9.0,py==1.8.0,pytest==4.4.0,six==1.12.0,stratego==0.0.0,sty==1.0.0b9
py36 run-test-pre: PYTHONHASHSEED='591797669'
py36 run-test: commands[0] | pytest
============================================================================================================== test session starts ==============================================================================================================
platform darwin -- Python 3.6.5, pytest-4.4.0, py-1.8.0, pluggy-0.9.0
cachedir: .tox/py36/.pytest_cache
rootdir: /Users/zayd/repos/pytest_bug, inifile: tox.ini
collected 0 items / 1 errors                                                                                                                                                                                                                    

==================================================================================================================== ERRORS =====================================================================================================================
_____________________________________________________________________________________________________ ERROR collecting stratego/printer.py ______________________________________________________________________________________________________
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/doctest.py:933: in find
    self._find(tests, obj, name, module, source_lines, globs, {})
.tox/py36/lib/python3.6/site-packages/_pytest/doctest.py:406: in _find
    self, tests, obj, name, module, source_lines, globs, seen
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/doctest.py:992: in _find
    if ((inspect.isroutine(inspect.unwrap(val))
.tox/py36/lib/python3.6/site-packages/_pytest/doctest.py:377: in _mock_aware_unwrap
    return real_unwrap(obj, stop=_is_mocked)
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/inspect.py:512: in unwrap
    while _is_wrapper(func):
../../.pyenv/versions/3.6.5/Python.framework/Versions/3.6/lib/python3.6/inspect.py:506: in _is_wrapper
    return hasattr(f, '__wrapped__') and not stop(f)
E   KeyError: '__wrapped__'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================ 1 error in 0.25 seconds ============================================================================================================
ERROR: InvocationError for command /Users/zayd/repos/pytest_bug/.tox/py36/bin/pytest (exited with code 2)
____________________________________________________________________________________________________________________ summary ____________________________________________________________________________________________________________________
ERROR:   py36: commands failed
```

My apologies if this bug is obvious or I am doing something wrong.

--- a/changelog/5404.bugfix.rst
+++ b/changelog/5404.bugfix.rst
@@ -1,0 +1,2 @@
+Emit a warning when attempting to unwrap a broken object raises an exception,
+for easier debugging (`#5080 <https://github.com/pytest-dev/pytest/issues/5080>`__).

--- a/src/_pytest/doctest.py
+++ b/src/_pytest/doctest.py
@@ -3,6 +3,7 @@ import inspect
 import platform
 import sys
 import traceback
+import warnings
 from contextlib import contextmanager
 
 import pytest
@@ -12,6 +13,7 @@ from _pytest._code.code import TerminalRepr
 from _pytest.compat import safe_getattr
 from _pytest.fixtures import FixtureRequest
 from _pytest.outcomes import Skipped
+from _pytest.warning_types import PytestWarning
 
 DOCTEST_REPORT_CHOICE_NONE = "none"
 DOCTEST_REPORT_CHOICE_CDIFF = "cdiff"
@@ -368,10 +370,18 @@ def _patch_unwrap_mock_aware():
     else:
 
         def _mock_aware_unwrap(obj, stop=None):
-            if stop is None:
-                return real_unwrap(obj, stop=_is_mocked)
-            else:
+            try:
+                if stop is None or stop is _is_mocked:
+                    return real_unwrap(obj, stop=_is_mocked)
                 return real_unwrap(obj, stop=lambda obj: _is_mocked(obj) or stop(obj))
+            except Exception as e:
+                warnings.warn(
+                    "Got %r when unwrapping %r.  This is usually caused "
+                    "by a violation of Python's object protocol; see e.g. "
+                    "https://github.com/pytest-dev/pytest/issues/5080" % (e, obj),
+                    PytestWarning,
+                )
+                raise
 
         inspect.unwrap = _mock_aware_unwrap
         try:

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -1,7 +1,10 @@
+import inspect
 import textwrap
 
 import pytest
 from _pytest.compat import MODULE_NOT_FOUND_ERROR
+from _pytest.doctest import _is_mocked
+from _pytest.doctest import _patch_unwrap_mock_aware
 from _pytest.doctest import DoctestItem
 from _pytest.doctest import DoctestModule
 from _pytest.doctest import DoctestTextfile
@@ -1224,3 +1227,24 @@ def test_doctest_mock_objects_dont_recurse_missbehaved(mock_module, testdir):
     )
     result = testdir.runpytest("--doctest-modules")
     result.stdout.fnmatch_lines(["* 1 passed *"])
+
+
+class Broken:
+    def __getattr__(self, _):
+        raise KeyError("This should be an AttributeError")
+
+
+@pytest.mark.parametrize(  # pragma: no branch (lambdas are not called)
+    "stop", [None, _is_mocked, lambda f: None, lambda f: False, lambda f: True]
+)
+def test_warning_on_unwrap_of_broken_object(stop):
+    bad_instance = Broken()
+    assert inspect.unwrap.__module__ == "inspect"
+    with _patch_unwrap_mock_aware():
+        assert inspect.unwrap.__module__ != "inspect"
+        with pytest.warns(
+            pytest.PytestWarning, match="^Got KeyError.* when unwrapping"
+        ):
+            with pytest.raises(KeyError):
+                inspect.unwrap(bad_instance, stop=stop)
+    assert inspect.unwrap.__module__ == "inspect"

